zsh-history-substring-search.zsh

```diff
@@ -233,20 +234,22 @@ _history-substring-search-begin() {
     _history_substring_search_query=$BUFFER
 
     #
-    # $BUFFER contains the text that is in the command-line currently.
-    # we put an extra "\\" before meta characters such as "\(" and "\)",
-    # so that they become "\\\(" and "\\\)".
-    #
-    local escaped_query=${BUFFER//(#m)[\][()|\\*?#<>~^]/\\$MATCH}
-
-    #
     # Find all occurrences of the search query in the history file.
     #
-    # (k) returns the "keys" (history index numbers) instead of the values
-    # (R) returns values in reverse older, so the index of the youngest
-    # matching history entry is at the head of the list.
-    #
-    _history_substring_search_raw_matches=(${(k)history[(R)(#$HISTORY_SUBSTRING_SEARCH_GLOBBING_FLAGS)*${escaped_query}*]})
+
+    local unified_query_parts=(${(L)=_history_substring_search_query});
+    local escaped_unified_query_parts=(${(q)unified_query_parts[@]})
+
+    local awk_query
+    if [ "$HISTORY_SUBSTRING_SEARCH_METHOD" = 'given_order' ]; then
+      awk_query="/${(j:.*:)escaped_unified_query_parts}/"      # find history by parts in given order
+    elif [ "$HISTORY_SUBSTRING_SEARCH_METHOD" = 'any_order' ]; then
+      awk_query="/${(j:/ && /:)escaped_unified_query_parts}/"  # find history by parts in any order
+    else
+      awk_query="/${(j: :)escaped_unified_query_parts}/"       # find history by parts as query
+    fi
+
+    _history_substring_search_raw_matches=($(history 0 | sort -r | awk "tolower(\$0) ~ $awk_query" | awk '{ print $1}'))
```

Comment:
sofar this is what I coded in pure zsh
`
local history_substring_search_raw_matches_string
    local history_command_id
    history -r 0 | while read history_command ; do
      if [[ TODO ]]
        history_command_id=${${(z)history_command}[1]}
        history_substring_search_raw_matches_string+=" "$history_command_id
      fi
    done
    _history_substring_search_raw_matches=($history_substring_search_raw_matches_string)
    echo $_history_substring_search_raw_matches
`